AY2021S1-CS2103-T16-3/tp
# Add archive command

### What this does
This PR adds an `archive` command to ResiReg. It archives the allocations (and perhaps bills in the future), but persists the students and rooms, as well as updates the application to work on a new semester. The successor of the semester should be properly calculated, meaning AY2020 Semester 2's successor should be AY2021 Semester 1.

I still need to do the following:
- [x] Display semester data from the semester in `MainWindow`
- [x] Implement the archival feature
- [x] Test `JsonAdaptedSemester` more effectively 
- [x] (EDIT: actually, test a bunch of stuff more effectively)


Closes #11.

### How to test

1. Launch the application.
2. Type `archive` in the command box and press <kbd>Enter</kbd>.
3. Observe that a folder denoting the previous semester is created, containing an `archive.json`, and the current semester's allocations are reset. The rooms and students however persist.

### Notes

I have created a `roomFees` map in my `Semester` model, which may be used in the future for billings. I've also written tests for those. However, this architecture may not be the most thought out, so whoever is working on billing in the future can feel free to modify it, and perhaps extract it out to its own class.

The file it saves to is `archive.json`, inside a folder representing the semester e.g. `AY2020S1`. This naming is up for debate, if you feel that it should change, please indicate in the review.

The semester in the GUI doesn't update. I'll need to make it `Observable` if it's required. Let me know in the review if you think it's necessary.

Fixes #104
